File: app/src/main/java/com/mrgames13/jimdo/feinstaubapp/RecyclerViewAdapters/SelectSensorAdapter.java
/*
 * Copyright © 2019 Marc Auberer. All rights reserved.
 */

package com.mrgames13.jimdo.feinstaubapp.RecyclerViewAdapters;

import android.content.Context;
import android.content.res.Resources;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.mrgames13.jimdo.feinstaubapp.CommonObjects.Sensor;
import com.mrgames13.jimdo.feinstaubapp.R;
import com.mrgames13.jimdo.feinstaubapp.Utils.StorageUtils;

import java.util.ArrayList;

import eu.davidea.flipview.FlipView;

public class SelectSensorAdapter extends RecyclerView.Adapter<SelectSensorAdapter.ViewHolder> {

    // Constants
    public static final int MODE_SELECTION_SINGLE = 10001;
    private static final int MODE_SELECTION_MULTI = 10002;

    // Utils packages
    private StorageUtils su;

    // Variables as objects
    private Resources res;
    private Handler h;
    private ArrayList<Sensor> sensors;
    private Sensor selected_sensor = null;
    private ArrayList<Sensor> selected_sensors = new ArrayList<>();
    private ViewHolder selected_sensor_holder;

    // Variables
    private int selection_mode;

    public SelectSensorAdapter(Context context, StorageUtils su, ArrayList<Sensor> sensors, int selection_mode) {
        this.su = su;
        this.res = context.getResources();
        this.h = new Handler();
        this.sensors = sensors;
        this.selection_mode = selection_mode;
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        // Variables as objects
        private FlipView item_icon;
        private TextView item_name;
        private TextView item_id;

        ViewHolder(View itemView) {
            super(itemView);
            // Initialize UI components
            item_icon = itemView.findViewById(R.id.item_icon);
            item_name = itemView.findViewById(R.id.item_name);
            item_id = itemView.findViewById(R.id.item_id);
            itemView.findViewById(R.id.item_more).setVisibility(View.GONE);
        }

        private void deselect() {
            item_icon.flip(false);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_sensor, parent, false);
        return new SelectSensorAdapter.ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull final ViewHolder h, int pos) {
        // Fill in data
        final Sensor sensor = sensors.get(pos);

        h.item_icon.getFrontLayout().getBackground().setColorFilter(sensor.getColor(), android.graphics.PorterDuff.Mode.SRC_IN);
        h.item_name.setText(sensor.getName());
        h.item_id.setText(res.getString(R.string.chip_id) + " " + sensor.getChipID());

        h.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                h.item_icon.flip(!h.item_icon.isFlipped());
            }
        });
        h.itemView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                h.item_icon.flip(!h.item_icon.isFlipped());
                return true;
            }
        });
        h.item_icon.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                h.item_icon.flip(!h.item_icon.isFlipped());
            }
        });
        h.item_icon.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                h.item_icon.flip(!h.item_icon.isFlipped());
                return true;
            }
        });
        h.item_icon.setOnFlippingListener(new FlipView.OnFlippingListener() {
            @Override
            public void onFlipped(FlipView flipView, boolean checked) {
                if(checked) {
                    if(selection_mode == MODE_SELECTION_SINGLE) {
                        SelectSensorAdapter.this.h.postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                if(selected_sensor_holder != null) selected_sensor_holder.deselect();
                                selected_sensor_holder = h;
                            }
                        }, 50);
                        selected_sensor = sensor;
                    } else if(selection_mode == MODE_SELECTION_MULTI) {
                        selected_sensors.add(sensor);
                    }
                } else {
                    if(selection_mode == MODE_SELECTION_SINGLE) {
                        if(selected_sensor.getChipID().equals(sensor.getChipID())) {
                            selected_sensor = null;
                            selected_sensor_holder = null;
                        }
                    } else if(selection_mode == MODE_SELECTION_MULTI) {
                        selected_sensors.remove(sensor);
                    }
                }
                h.itemView.setBackgroundColor(res.getColor(checked ? R.color.color_selection : R.color.transparent));
            }
        });

        h.itemView.findViewById(R.id.item_own_sensor).setVisibility(su.isSensorExisting(sensor.getChipID()) ? View.VISIBLE : View.GONE);
    }

    @Override
    public int getItemCount() {
        return sensors.size();
    }

    public Sensor getSelectedSensor() {
        return selected_sensor;
    }
}

File: app/src/main/java/com/mrgames13/jimdo/feinstaubapp/RecyclerViewAdapters/SelectSensorAdapter.kt
/*
 * Copyright © 2019 Marc Auberer. All rights reserved.
 */

package com.mrgames13.jimdo.feinstaubapp.RecyclerViewAdapters

import android.content.Context
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.core.graphics.BlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import androidx.recyclerview.widget.RecyclerView
import com.mrgames13.jimdo.feinstaubapp.CommonObjects.Sensor
import com.mrgames13.jimdo.feinstaubapp.R
import com.mrgames13.jimdo.feinstaubapp.Utils.StorageUtils
import kotlinx.android.synthetic.main.item_sensor.view.*
import java.util.*

class SelectSensorAdapter(private var context: Context, private val su: StorageUtils, private val sensors: ArrayList<Sensor>, private val selection_mode: Int) : RecyclerView.Adapter<SelectSensorAdapter.ViewHolder>() {

    // Variables as objects
    private val h: Handler = Handler()
    internal var selectedSensor: Sensor? = null
    private val selectedSensors = ArrayList<Sensor>()
    private var selectedSensorHolder: ViewHolder? = null

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            // Initialize UI components
            itemView.item_more.visibility = View.GONE
        }

        internal fun deselect() {
            itemView.item_icon.flip(false)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.item_sensor, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(h: ViewHolder, pos: Int) {
        // Fill in data
        val sensor = sensors[pos]

        h.itemView.item_icon.frontLayout.background.colorFilter = BlendModeColorFilterCompat.createBlendModeColorFilterCompat(sensor.color, BlendModeCompat.SRC_IN)
        h.itemView.item_name.text = sensor.name
        h.itemView.item_id.text = context.getString(R.string.chip_id) + " " + sensor.chipID

        h.itemView.setOnClickListener { h.itemView.item_icon.flip(!h.itemView.item_icon.isFlipped) }
        h.itemView.setOnLongClickListener {
            h.itemView.item_icon.flip(!h.itemView.item_icon.isFlipped)
            true
        }
        h.itemView.item_icon.setOnClickListener { h.itemView.item_icon.flip(!h.itemView.item_icon.isFlipped) }
        h.itemView.item_icon.setOnLongClickListener {
            h.itemView.item_icon.flip(!h.itemView.item_icon.isFlipped)
            true
        }
        h.itemView.item_icon.setOnFlippingListener { _, checked ->
            if (checked) {
                if (selection_mode == MODE_SELECTION_SINGLE) {
                    this@SelectSensorAdapter.h.postDelayed({
                        selectedSensorHolder?.deselect()
                        selectedSensorHolder = h
                    }, 50)
                    selectedSensor = sensor
                } else if (selection_mode == MODE_SELECTION_MULTI) {
                    selectedSensors.add(sensor)
                }
            } else {
                if (selection_mode == MODE_SELECTION_SINGLE) {
                    if (selectedSensor?.chipID == sensor.chipID) {
                        selectedSensor = null
                        selectedSensorHolder = null
                    }
                } else if (selection_mode == MODE_SELECTION_MULTI) {
                    selectedSensors.remove(sensor)
                }
            }
            h.itemView.setBackgroundColor(ContextCompat.getColor(context, if (checked) R.color.color_selection else R.color.transparent))
        }

        h.itemView.item_own_sensor.visibility = if (su.isSensorExisting(sensor.chipID)) View.VISIBLE else View.GONE
    }

    override fun getItemCount(): Int {
        return sensors.size
    }

    companion object {
        // Constants
        const val MODE_SELECTION_SINGLE = 10001
        private const val MODE_SELECTION_MULTI = 10002
    }
}